src/macros.c

```diff
@@ -0,0 +1,342 @@
+	POST_MACRO,		// after {{macro}}
+};
+
+// Text chunks for text transform.
+typedef struct{
+	int len;
+	char *with;
+} txt_chunk;
+// }}}
+
+// Hash Table for Macros {{{
+// MurMur3 hash function
+// Adapted from:
+// https://github.com/stephane-martin/cycapture/blob/master/cycapture/murmur.c
+// Credit: Stephane Martin
+static FORCE_INLINE uint32_t hash(const void *data, size_t nbytes)
+{
+	if (data == NULL || nbytes == 0)
+		return 0;
+
+	const uint32_t c1 = 0xcc9e2d51;
+	const uint32_t c2 = 0x1b873593;
+
+	const int nblocks = nbytes / 4;
+	const uint32_t *blocks = (const uint32_t *) (data);
+	const uint8_t *tail = (const uint8_t *) data + nblocks * 4;
+
+	uint32_t h = 0;
+
+	int i;
+	uint32_t k;
+	for (i = 0; i < nblocks; i++) {
+		k = blocks[i];
+
+		k *= c1;
+		k = (k << 15) | (k >> (32 - 15));
+		k *= c2;
+
+		h ^= k;
+		h = (h << 13) | (h >> (32 - 13));
+		h = (h * 5) + 0xe6546b64;
+	}
+
+	k = 0;
+	switch (nbytes & 3) {
+		case 3:
+			k ^= tail[2] << 16;
+			/* FALLTHRU */
+		case 2:
+			k ^= tail[1] << 8;
+			/* FALLTHRU */
+		case 1:
+			k ^= tail[0];
+			k *= c1;
+			k = (k << 15) | (k >> (32 - 15));
+			k *= c2;
+			h ^= k;
+	};
+
+	h ^= nbytes;
+
+	h ^= h >> 16;
+	h *= 0x85ebca6b;
+	h ^= h >> 13;
+	h *= 0xc2b2ae35;
+	h ^= h >> 16;
+
+	return h;
+}
+
+static FORCE_INLINE uint32_t hash_to_index(const ht_t *ht, uint32_t h)
+{
+	return h % ht->size;
+}
+
+// Trim a string left and right and return its new length
+static FORCE_INLINE size_t str_trim(char *src)
+{
+	size_t i = 0;
+	size_t ts = 0;
+	size_t j = safe_strlen(src) - 1;
+	if(src == NULL || j <= 0)
+		return 0;
+	while(j > 0 && isspace(src[j])) j--;
+	while(i <= j && isspace(src[i])) i++;
+	ts = j - i + 1;
+	memcpy(src, src+i, ts);
+	src[ts] = '\0';
+	return ts;
+}
+
+ht_t *ht_create(const size_t size)
+{
+	ht_t *ht = malloc(sizeof(ht_t));
+	if(ht == NULL)
+		return ht;
+	ht->size = size;
+	ht->entries = calloc(size, sizeof(macro_t));
+	if(ht->entries == NULL){
+		free(ht);
+		return NULL;
+	}
+	return ht;
+}
+
+void ht_set(const ht_t *ht, const void *key, const size_t k_len, const void *value, const size_t v_len)
+{
+	uint32_t h = 0;
+	size_t slot = 0;
+	size_t i = 0;
+
+	if(ht == NULL || key == NULL || k_len == 0)
+		return;
+
+	h = hash(key, k_len);
+
+	for (i = 0; i < ht->size; i++){
+		slot = hash_to_index(ht, (h+1));
+		if(ht->entries[slot].key == NULL){
+			ht->entries[slot].key = malloc(k_len);
+			ht->entries[slot].value = malloc(v_len);
+			memcpy(ht->entries[slot].key, key, k_len);
+			memcpy(ht->entries[slot].value, value, v_len);
+			ht->entries[slot].k_len = k_len;
+			ht->entries[slot].v_len = v_len;
+			break;
+		} else if(k_len == ht->entries[slot].k_len){
+			if(memcmp(ht->entries[slot].key, key, k_len) == 0){
+				free(ht->entries[slot].value);
+				ht->entries[slot].value = malloc(v_len);
+				memcpy(ht->entries[slot].value, value, v_len);
+				ht->entries[slot].k_len = k_len;
+				ht->entries[slot].v_len = v_len;
+				break;
+			}
+		}
+	}
```

Comment:
What is this code even doing? It's iterating `ht->size` times always making the same checks: `slot` has the same value every iteration, `k_len` is `const`, no value of `ht` should change. ???

Was that `(h+1)` supposed to be an `h++`? Even if that's what's wrong, this doesn't look like an hash table to me!

What is this data structure supposed to do when there is an hash collision?